.github/workflows/test.yml

```diff
@@ -31,54 +41,96 @@ jobs:
           is_base_os_version: true
     steps:
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 30
       - uses: actions/setup-node@v3
         with:
           node-version: ${{ matrix.node }}
 
+      - name: env
+        run: |
+          echo 'dist-restore-keys<<EOF' >> $GITHUB_ENV
+          echo "$(git --no-pager log -9 --skip 1 --no-merges --pretty=format:'%H--test-dist-${{ matrix.node }}-${{ matrix.os }}')" >> $GITHUB_ENV
+          echo 'EOF' >> $GITHUB_ENV
+          echo "commit-sha=$(git --no-pager log -1 --no-merges --pretty=format:'%H')" >> $GITHUB_ENV
+          echo "build-and-test-all-packages=$(echo ${{ github.event.inputs.run_index }})"
+
+      - name: fetch previous build artifacts
+        uses: actions/cache@v3.0.3
+        with:
+          path: |
+            .cached-commit
+            cli/*/dist/**
+            experimental/*/dist/**
+            packages/*/dist/**
+            plugin-packs/*/dist/**
+            plugins/*/dist/**
+          key: ${{ env.commit-sha }}--test-dist-${{ matrix.node }}-${{ matrix.os }}
+          restore-keys: ${{ env.dist-restore-keys }}
+
         # Build, lint and PostCSS Tape tests must all work and pass :
         # - with exact dependencies from package-lock.json
         # - without requiring postinstall scripts from dependencies to run
-      - run: npm ci --ignore-scripts
-      - run: npm run build
-      - run: npm run lint
+      - name: build
+        run: |
+          npm ci --ignore-scripts
+          VERBOSE=true node './.github/bin/modified-workspaces/log-modified-workspaces.mjs'
+          npm run build --if-present $(node './.github/bin/modified-workspaces/log-modified-workspaces.mjs')
+
+      - name: lint
+        run: npm run lint --if-present $(node './.github/bin/modified-workspaces/log-modified-workspaces.mjs')
         if: matrix.is_base_node_version && matrix.is_base_os_version
 
-        # PostCSS Tape tests
-      - run: npm run test
+        # Basic tests
+      - name: test
+        run: npm run test --if-present $(node './.github/bin/modified-workspaces/log-modified-workspaces.mjs')
         env:
           ENABLE_ANNOTATIONS_FOR_NODE: ${{ matrix.is_base_node_version }}
           ENABLE_ANNOTATIONS_FOR_OS: ${{ matrix.is_base_os_version }}
 
         # CLI Tests
         # running "npm install" places the cli in node_modules/bin
-      - run: npm install --ignore-scripts
-      - run: npm run test:cli --workspace="@csstools/postcss-base-plugin"
+      - name: test:cli
+        run: |
+          npm install --ignore-scripts
+          npm run test:cli --if-present $(node './.github/bin/modified-workspaces/log-modified-workspaces.mjs')
 
-  # Run e2e tests
-  test_e2e:
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v3
```

Comment:
Everything now runs in a single contiguous job so that we minimise how much building we need to do.